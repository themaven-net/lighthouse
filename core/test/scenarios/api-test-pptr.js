/**
 * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

import jestMock from 'jest-mock';

import * as api from '../../index.js';
import {createTestState, getAuditsBreakdown} from './pptr-test-utils.js';
import {LH_ROOT} from '../../../root.js';

describe('Fraggle Rock API', function() {
  // eslint-disable-next-line no-invalid-this
  this.timeout(120_000);

  const state = createTestState();

  state.installSetupAndTeardownHooks();

  async function setupTestPage() {
    await state.page.goto(`${state.serverBaseUrl}/onclick.html`, {timeout: 90_000});
    // Wait for the javascript to run.
    await state.page.waitForSelector('button');
    await state.page.click('button');
    // Wait for the violations to appear (and console to be populated).
    await state.page.waitForSelector('input');
  }

  describe('snapshot', () => {
    beforeEach(() => {
      const {server} = state;
      server.baseDir = `${LH_ROOT}/core/test/fixtures/fraggle-rock/snapshot-basic`;
    });

    it('should compute accessibility results on the page as-is', async () => {
      await setupTestPage();

      const result = await api.snapshot(state.page);
      if (!result) throw new Error('Lighthouse failed to produce a result');

      const {lhr, artifacts} = result;
      const url = `${state.serverBaseUrl}/onclick.html#done`;
      expect(artifacts.URL).toEqual({
        finalDisplayedUrl: url,
      });

      const accessibility = lhr.categories.accessibility;
      expect(accessibility.score).toBeLessThan(1);

      const {auditResults, erroredAudits, failedAudits} = getAuditsBreakdown(lhr);
      expect(auditResults.map(audit => audit.id).sort()).toMatchSnapshot();

      expect(erroredAudits).toHaveLength(0);
      expect(failedAudits.map(audit => audit.id)).toContain('label');
    });
  });

  describe('startTimespan', () => {
    beforeEach(() => {
      const {server} = state;
      server.baseDir = `${LH_ROOT}/core/test/fixtures/fraggle-rock/snapshot-basic`;
    });

    it('should compute ConsoleMessage results across a span of time', async () => {
      const run = await api.startTimespan(state.page);

      await setupTestPage();

      // Wait long enough to ensure a paint after button interaction.
      await state.page.waitForTimeout(200);

      const result = await run.endTimespan();
      if (!result) throw new Error('Lighthouse failed to produce a result');

      const {lhr, artifacts} = result;
      expect(artifacts.URL).toEqual({
        finalDisplayedUrl: `${state.serverBaseUrl}/onclick.html#done`,
      });

      const bestPractices = lhr.categories['best-practices'];
      expect(bestPractices.score).toBeLessThan(1);

      const {
        auditResults,
        erroredAudits,
        failedAudits,
        notApplicableAudits,
      } = getAuditsBreakdown(lhr);
      expect(auditResults.map(audit => audit.id).sort()).toMatchSnapshot();

      expect(notApplicableAudits.map(audit => audit.id).sort()).toMatchSnapshot();
      expect(notApplicableAudits.map(audit => audit.id)).not.toContain('total-blocking-time');

      expect(erroredAudits).toHaveLength(0);
      expect(failedAudits.map(audit => audit.id)).toContain('errors-in-console');

      const errorsInConsole = lhr.audits['errors-in-console'];
      if (!errorsInConsole.details) throw new Error('Error in consoles audit missing details');
      if (errorsInConsole.details.type !== 'table') throw new Error('Unexpected details');
      const errorLogs = errorsInConsole.details.items;
      const matchingLog = errorLogs.find(
        log =>
          log.source === 'console.error' &&
          String(log.description || '').includes('violations added')
      );
      // If we couldn't find it, assert something similar on the object that we know will fail
      // for a better debug message.
      if (!matchingLog) expect(errorLogs).toContain({description: /violations added/});

      // Check that network request information was computed.
      expect(lhr.audits).toHaveProperty('total-byte-weight');
      const details = lhr.audits['total-byte-weight'].details;
      if (!details || details.type !== 'table') throw new Error('Unexpected byte weight details');
      expect(details.items).toMatchObject([{url: `${state.serverBaseUrl}/onclick.html`}]);
    });

    it('should compute results from timespan after page load', async () => {
      const {page, serverBaseUrl} = state;
      await page.goto(`${serverBaseUrl}/onclick.html`);
      await page.waitForSelector('button');

      const run = await api.startTimespan(state.page);

      await page.click('button');
      await page.waitForSelector('input');

      // Wait long enough to ensure a paint after button interaction.
      await page.waitForTimeout(200);

      const result = await run.endTimespan();

      if (!result) throw new Error('Lighthouse failed to produce a result');

      expect(result.artifacts.URL).toEqual({
        finalDisplayedUrl: `${serverBaseUrl}/onclick.html#done`,
      });

      const {auditResults, erroredAudits, notApplicableAudits} = getAuditsBreakdown(result.lhr);
      expect(auditResults.map(audit => audit.id).sort()).toMatchSnapshot();

      expect(notApplicableAudits.map(audit => audit.id).sort()).toMatchSnapshot();
      expect(notApplicableAudits.map(audit => audit.id)).not.toContain('total-blocking-time');

      expect(erroredAudits).toHaveLength(0);
    });

    // eslint-disable-next-line max-len
    it('should know target type of network requests from frames created before timespan', async () => {
      state.server.baseDir = `${LH_ROOT}/cli/test/fixtures`;
      const {page, serverBaseUrl} = state;

      await page.goto(`${serverBaseUrl}/oopif-scripts-timespan.html`);

      const run = await api.startTimespan(state.page);
      for (const iframe of page.frames()) {
        if (iframe.url().includes('/oopif-simple-page.html')) {
          iframe.click('button');
        }
      }
      await page.waitForNetworkIdle().catch(() => {});
      const result = await run.endTimespan();

      if (!result) throw new Error('Lighthouse failed to produce a result');

      const networkRequestsDetails = /** @type {LH.Audit.Details.Table} */ (
        result.lhr.audits['network-requests'].details);
      const networkRequests = networkRequestsDetails?.items
        .map((r) => ({url: r.url, sessionTargetType: r.sessionTargetType}))
        // @ts-expect-error
        .sort((a, b) => a.url.localeCompare(b.url));
      expect(networkRequests).toHaveLength(4);
      expect(networkRequests.filter(r => r.sessionTargetType === 'page')).toHaveLength(2);
      expect(networkRequests.filter(r => r.sessionTargetType === 'iframe')).toHaveLength(2);
      expect(networkRequests).toMatchInlineSnapshot(`
Array [
  Object {
    "sessionTargetType": "page",
    "url": "http://localhost:10200/simple-script.js",
  },
  Object {
    "sessionTargetType": "page",
    "url": "http://localhost:10200/simple-worker.js",
  },
  Object {
    "sessionTargetType": "iframe",
    "url": "http://localhost:10503/simple-script.js",
  },
  Object {
    "sessionTargetType": "iframe",
    "url": "http://localhost:10503/simple-worker.js",
  },
]
`);
    });
  });

  describe('navigation', () => {
    beforeEach(() => {
      const {server} = state;
      server.baseDir = `${LH_ROOT}/core/test/fixtures/fraggle-rock/navigation-basic`;
    });

    it('should compute both snapshot & timespan results', async () => {
      const {page, serverBaseUrl} = state;
      const url = `${serverBaseUrl}/index.html`;
      const result = await api.navigation(page, url);
      if (!result) throw new Error('Lighthouse failed to produce a result');

      const {lhr, artifacts} = result;
      expect(artifacts.URL).toEqual({
        requestedUrl: url,
        mainDocumentUrl: url,
        finalDisplayedUrl: url,
      });

      const {auditResults, failedAudits, erroredAudits} = getAuditsBreakdown(lhr);
      expect(auditResults.map(audit => audit.id).sort()).toMatchSnapshot();
      expect(erroredAudits).toHaveLength(0);

      const failedAuditIds = failedAudits.map(audit => audit.id);
      expect(failedAuditIds).toContain('label');
      expect(failedAuditIds).toContain('errors-in-console');

      // Check that network request information was computed.
      expect(lhr.audits).toHaveProperty('total-byte-weight');
      const details = lhr.audits['total-byte-weight'].details;
      if (!details || details.type !== 'table') throw new Error('Unexpected byte weight details');
      expect(details.items).toMatchObject([{url}]);

      // Check that performance metrics were computed.
      expect(lhr.audits).toHaveProperty('first-contentful-paint');
      expect(Number.isFinite(lhr.audits['first-contentful-paint'].numericValue)).toBe(true);
    });

    it('should compute results with callback requestor', async () => {
      const {page, serverBaseUrl} = state;
      const requestedUrl = `${serverBaseUrl}/?redirect=/index.html`;
      const mainDocumentUrl = `${serverBaseUrl}/index.html`;
      await page.goto(`${serverBaseUrl}/links-to-index.html`);

      const requestor = jestMock.fn(async () => {
        await page.click('a');
      });

      const result = await api.navigation(page, requestor);
      if (!result) throw new Error('Lighthouse failed to produce a result');

      expect(requestor).toHaveBeenCalled();

      const {lhr, artifacts} = result;
      expect(lhr.requestedUrl).toEqual(requestedUrl);
      expect(lhr.finalDisplayedUrl).toEqual(mainDocumentUrl);
      expect(artifacts.URL).toEqual({
        requestedUrl,
        mainDocumentUrl,
        finalDisplayedUrl: mainDocumentUrl,
      });

      const {auditResults, failedAudits, erroredAudits} = getAuditsBreakdown(lhr);
      expect(auditResults.map(audit => audit.id).sort()).toMatchSnapshot();
      expect(erroredAudits).toHaveLength(0);

      const failedAuditIds = failedAudits.map(audit => audit.id);
      expect(failedAuditIds).toContain('label');
      expect(failedAuditIds).toContain('errors-in-console');

      // Check that network request information was computed.
      expect(lhr.audits).toHaveProperty('total-byte-weight');
      const details = lhr.audits['total-byte-weight'].details;
      if (!details || details.type !== 'table') throw new Error('Unexpected byte weight details');
      expect(details.items).toMatchObject([
        {url: mainDocumentUrl},
        {url: `${serverBaseUrl}/?redirect=/index.html`},
      ]);

      // Check that performance metrics were computed.
      expect(lhr.audits).toHaveProperty('first-contentful-paint');
      expect(Number.isFinite(lhr.audits['first-contentful-paint'].numericValue)).toBe(true);
    });
  });
});
