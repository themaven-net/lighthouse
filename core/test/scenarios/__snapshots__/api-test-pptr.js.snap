// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fraggle Rock API navigation should compute both snapshot & timespan results 1`] = `
Array [
  "accesskeys",
  "aria-allowed-attr",
  "aria-command-name",
  "aria-hidden-body",
  "aria-hidden-focus",
  "aria-input-field-name",
  "aria-meter-name",
  "aria-progressbar-name",
  "aria-required-attr",
  "aria-required-children",
  "aria-required-parent",
  "aria-roles",
  "aria-toggle-field-name",
  "aria-tooltip-name",
  "aria-treeitem-name",
  "aria-valid-attr",
  "aria-valid-attr-value",
  "bf-cache",
  "bootup-time",
  "button-name",
  "bypass",
  "canonical",
  "charset",
  "color-contrast",
  "content-width",
  "crawlable-anchors",
  "critical-request-chains",
  "csp-xss",
  "cumulative-layout-shift",
  "custom-controls-labels",
  "custom-controls-roles",
  "definition-list",
  "deprecations",
  "diagnostics",
  "dlitem",
  "doctype",
  "document-title",
  "dom-size",
  "duplicate-id-active",
  "duplicate-id-aria",
  "duplicated-javascript",
  "efficient-animated-content",
  "errors-in-console",
  "final-screenshot",
  "first-contentful-paint",
  "first-meaningful-paint",
  "focus-traps",
  "focusable-controls",
  "font-display",
  "font-size",
  "form-field-multiple-labels",
  "frame-title",
  "geolocation-on-start",
  "heading-order",
  "hreflang",
  "html-has-lang",
  "html-lang-valid",
  "http-status-code",
  "image-alt",
  "image-aspect-ratio",
  "image-size-responsive",
  "input-image-alt",
  "inspector-issues",
  "installable-manifest",
  "interactive",
  "interactive-element-affordance",
  "is-crawlable",
  "is-on-https",
  "js-libraries",
  "label",
  "largest-contentful-paint",
  "largest-contentful-paint-element",
  "layout-shift-elements",
  "lcp-lazy-loaded",
  "legacy-javascript",
  "link-name",
  "link-text",
  "list",
  "listitem",
  "logical-tab-order",
  "long-tasks",
  "main-thread-tasks",
  "mainthread-work-breakdown",
  "managed-focus",
  "maskable-icon",
  "max-potential-fid",
  "meta-description",
  "meta-refresh",
  "meta-viewport",
  "metrics",
  "modern-image-formats",
  "network-requests",
  "network-rtt",
  "network-server-latency",
  "no-document-write",
  "no-unload-listeners",
  "non-composited-animations",
  "notification-on-start",
  "object-alt",
  "offscreen-content-hidden",
  "offscreen-images",
  "paste-preventing-inputs",
  "performance-budget",
  "plugins",
  "preload-fonts",
  "prioritize-lcp-image",
  "pwa-cross-browser",
  "pwa-each-page-has-url",
  "pwa-page-transitions",
  "redirects",
  "render-blocking-resources",
  "resource-summary",
  "robots-txt",
  "screenshot-thumbnails",
  "script-treemap-data",
  "server-response-time",
  "service-worker",
  "speed-index",
  "splash-screen",
  "structured-data",
  "tabindex",
  "tap-targets",
  "td-headers-attr",
  "th-has-data-cells",
  "themed-omnibox",
  "third-party-facades",
  "third-party-summary",
  "timing-budget",
  "total-blocking-time",
  "total-byte-weight",
  "unminified-css",
  "unminified-javascript",
  "unsized-images",
  "unused-css-rules",
  "unused-javascript",
  "use-landmarks",
  "user-timings",
  "uses-http2",
  "uses-long-cache-ttl",
  "uses-optimized-images",
  "uses-passive-event-listeners",
  "uses-rel-preconnect",
  "uses-rel-preload",
  "uses-responsive-images",
  "uses-text-compression",
  "valid-lang",
  "valid-source-maps",
  "video-caption",
  "viewport",
  "visual-order-follows-dom",
]
`;

exports[`Fraggle Rock API navigation should compute results with callback requestor 1`] = `
Array [
  "accesskeys",
  "aria-allowed-attr",
  "aria-command-name",
  "aria-hidden-body",
  "aria-hidden-focus",
  "aria-input-field-name",
  "aria-meter-name",
  "aria-progressbar-name",
  "aria-required-attr",
  "aria-required-children",
  "aria-required-parent",
  "aria-roles",
  "aria-toggle-field-name",
  "aria-tooltip-name",
  "aria-treeitem-name",
  "aria-valid-attr",
  "aria-valid-attr-value",
  "bf-cache",
  "bootup-time",
  "button-name",
  "bypass",
  "canonical",
  "charset",
  "color-contrast",
  "content-width",
  "crawlable-anchors",
  "critical-request-chains",
  "csp-xss",
  "cumulative-layout-shift",
  "custom-controls-labels",
  "custom-controls-roles",
  "definition-list",
  "deprecations",
  "diagnostics",
  "dlitem",
  "doctype",
  "document-title",
  "dom-size",
  "duplicate-id-active",
  "duplicate-id-aria",
  "duplicated-javascript",
  "efficient-animated-content",
  "errors-in-console",
  "final-screenshot",
  "first-contentful-paint",
  "first-meaningful-paint",
  "focus-traps",
  "focusable-controls",
  "font-display",
  "font-size",
  "form-field-multiple-labels",
  "frame-title",
  "geolocation-on-start",
  "heading-order",
  "hreflang",
  "html-has-lang",
  "html-lang-valid",
  "http-status-code",
  "image-alt",
  "image-aspect-ratio",
  "image-size-responsive",
  "input-image-alt",
  "inspector-issues",
  "installable-manifest",
  "interactive",
  "interactive-element-affordance",
  "is-crawlable",
  "is-on-https",
  "js-libraries",
  "label",
  "largest-contentful-paint",
  "largest-contentful-paint-element",
  "layout-shift-elements",
  "lcp-lazy-loaded",
  "legacy-javascript",
  "link-name",
  "link-text",
  "list",
  "listitem",
  "logical-tab-order",
  "long-tasks",
  "main-thread-tasks",
  "mainthread-work-breakdown",
  "managed-focus",
  "maskable-icon",
  "max-potential-fid",
  "meta-description",
  "meta-refresh",
  "meta-viewport",
  "metrics",
  "modern-image-formats",
  "network-requests",
  "network-rtt",
  "network-server-latency",
  "no-document-write",
  "no-unload-listeners",
  "non-composited-animations",
  "notification-on-start",
  "object-alt",
  "offscreen-content-hidden",
  "offscreen-images",
  "paste-preventing-inputs",
  "performance-budget",
  "plugins",
  "preload-fonts",
  "prioritize-lcp-image",
  "pwa-cross-browser",
  "pwa-each-page-has-url",
  "pwa-page-transitions",
  "redirects",
  "render-blocking-resources",
  "resource-summary",
  "robots-txt",
  "screenshot-thumbnails",
  "script-treemap-data",
  "server-response-time",
  "service-worker",
  "speed-index",
  "splash-screen",
  "structured-data",
  "tabindex",
  "tap-targets",
  "td-headers-attr",
  "th-has-data-cells",
  "themed-omnibox",
  "third-party-facades",
  "third-party-summary",
  "timing-budget",
  "total-blocking-time",
  "total-byte-weight",
  "unminified-css",
  "unminified-javascript",
  "unsized-images",
  "unused-css-rules",
  "unused-javascript",
  "use-landmarks",
  "user-timings",
  "uses-http2",
  "uses-long-cache-ttl",
  "uses-optimized-images",
  "uses-passive-event-listeners",
  "uses-rel-preconnect",
  "uses-rel-preload",
  "uses-responsive-images",
  "uses-text-compression",
  "valid-lang",
  "valid-source-maps",
  "video-caption",
  "viewport",
  "visual-order-follows-dom",
]
`;

exports[`Fraggle Rock API snapshot should compute accessibility results on the page as-is 1`] = `
Array [
  "accesskeys",
  "aria-allowed-attr",
  "aria-command-name",
  "aria-hidden-body",
  "aria-hidden-focus",
  "aria-input-field-name",
  "aria-meter-name",
  "aria-progressbar-name",
  "aria-required-attr",
  "aria-required-children",
  "aria-required-parent",
  "aria-roles",
  "aria-toggle-field-name",
  "aria-tooltip-name",
  "aria-treeitem-name",
  "aria-valid-attr",
  "aria-valid-attr-value",
  "button-name",
  "bypass",
  "color-contrast",
  "crawlable-anchors",
  "custom-controls-labels",
  "custom-controls-roles",
  "definition-list",
  "dlitem",
  "doctype",
  "document-title",
  "dom-size",
  "duplicate-id-active",
  "duplicate-id-aria",
  "focus-traps",
  "focusable-controls",
  "font-size",
  "form-field-multiple-labels",
  "frame-title",
  "heading-order",
  "html-has-lang",
  "html-lang-valid",
  "image-alt",
  "image-aspect-ratio",
  "image-size-responsive",
  "input-image-alt",
  "interactive-element-affordance",
  "js-libraries",
  "label",
  "link-name",
  "link-text",
  "list",
  "listitem",
  "logical-tab-order",
  "managed-focus",
  "meta-description",
  "meta-refresh",
  "meta-viewport",
  "object-alt",
  "offscreen-content-hidden",
  "paste-preventing-inputs",
  "plugins",
  "robots-txt",
  "structured-data",
  "tabindex",
  "tap-targets",
  "td-headers-attr",
  "th-has-data-cells",
  "unsized-images",
  "use-landmarks",
  "uses-responsive-images-snapshot",
  "valid-lang",
  "video-caption",
  "viewport",
  "visual-order-follows-dom",
]
`;

exports[`Fraggle Rock API startTimespan should compute ConsoleMessage results across a span of time 1`] = `
Array [
  "bf-cache",
  "bootup-time",
  "cumulative-layout-shift",
  "deprecations",
  "duplicated-javascript",
  "efficient-animated-content",
  "errors-in-console",
  "experimental-interaction-to-next-paint",
  "final-screenshot",
  "image-aspect-ratio",
  "image-size-responsive",
  "inspector-issues",
  "is-on-https",
  "layout-shift-elements",
  "legacy-javascript",
  "long-tasks",
  "main-thread-tasks",
  "mainthread-work-breakdown",
  "modern-image-formats",
  "network-requests",
  "network-rtt",
  "network-server-latency",
  "no-document-write",
  "no-unload-listeners",
  "non-composited-animations",
  "preload-fonts",
  "resource-summary",
  "screenshot-thumbnails",
  "script-treemap-data",
  "third-party-summary",
  "total-blocking-time",
  "total-byte-weight",
  "unminified-css",
  "unminified-javascript",
  "unsized-images",
  "unused-css-rules",
  "unused-javascript",
  "user-timings",
  "uses-http2",
  "uses-long-cache-ttl",
  "uses-optimized-images",
  "uses-passive-event-listeners",
  "uses-responsive-images",
  "uses-text-compression",
  "valid-source-maps",
  "work-during-interaction",
]
`;

exports[`Fraggle Rock API startTimespan should compute ConsoleMessage results across a span of time 2`] = `
Array [
  "non-composited-animations",
  "preload-fonts",
  "third-party-summary",
  "unsized-images",
  "user-timings",
]
`;

exports[`Fraggle Rock API startTimespan should compute results from timespan after page load 1`] = `
Array [
  "bf-cache",
  "bootup-time",
  "cumulative-layout-shift",
  "deprecations",
  "duplicated-javascript",
  "efficient-animated-content",
  "errors-in-console",
  "experimental-interaction-to-next-paint",
  "final-screenshot",
  "image-aspect-ratio",
  "image-size-responsive",
  "inspector-issues",
  "is-on-https",
  "layout-shift-elements",
  "legacy-javascript",
  "long-tasks",
  "main-thread-tasks",
  "mainthread-work-breakdown",
  "modern-image-formats",
  "network-requests",
  "network-rtt",
  "network-server-latency",
  "no-document-write",
  "no-unload-listeners",
  "non-composited-animations",
  "preload-fonts",
  "resource-summary",
  "screenshot-thumbnails",
  "script-treemap-data",
  "third-party-summary",
  "total-blocking-time",
  "total-byte-weight",
  "unminified-css",
  "unminified-javascript",
  "unsized-images",
  "unused-css-rules",
  "unused-javascript",
  "user-timings",
  "uses-http2",
  "uses-long-cache-ttl",
  "uses-optimized-images",
  "uses-passive-event-listeners",
  "uses-responsive-images",
  "uses-text-compression",
  "valid-source-maps",
  "work-during-interaction",
]
`;

exports[`Fraggle Rock API startTimespan should compute results from timespan after page load 2`] = `
Array [
  "duplicated-javascript",
  "efficient-animated-content",
  "legacy-javascript",
  "modern-image-formats",
  "network-rtt",
  "network-server-latency",
  "non-composited-animations",
  "preload-fonts",
  "third-party-summary",
  "unminified-css",
  "unminified-javascript",
  "unsized-images",
  "unused-css-rules",
  "unused-javascript",
  "user-timings",
  "uses-optimized-images",
  "uses-responsive-images",
  "uses-text-compression",
]
`;
